### PR TITLE
gsoc26: multi circuit support in canonical.ts

### LINKED ISSUE #1069: (GSoC'26) Week 2: Implement Multi-Circuit Project Export and Import

### PR DESCRIPTION
**Pull request by Harkeerat24**
Fixes #1069 

 

#### Describe the changes you have made in this PR -
- Extends `canonical.ts` to handle a full CircuitVerse **project** (multiple independent scopes)

- **`CanonicalProject` type** — top-level envelope with a `formatVersion`, a per-circuit map keyed by scope ID, and a project-level `canonicalHash`.

- **`canonicaliseProject`** — accepts a single scope or an array of scopes, calls `canonicaliseScope` on each, collects the individual hashes, sorts them, and produces a project-level SHA-256. Hence, the combined hash is stable regardless of the order scopes appear in the input array.

### Screenshots of the UI changes (If any) -

No UI changes.
---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**Explain your implementation approach:**

- Here circuits are treated as fully independent: each scope is canonicalised in isolation and the results are combined by hashing the sorted set of scope hashes. 
- I deliberately kept this PR free of subcircuit dependency logic. 
- The next PR will have the subcircuit implementation.
- You can see the json in the console by using: `await (await import('/simulatorvue/v1/v1/src/simulator/src/data/canonical.ts')).canonicaliseProject(Object.values((await import('/simulatorvue/v1/v1/src/simulator/src/circuit.ts')).scopeList))`

image

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [ ] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

## Release Notes

* **New Features**
  * Added deterministic canonicalization for simulator circuit scopes and projects, producing stable identifiers and repeatable hashes across runs.
  * Canonicalization now normalizes circuit structure, connectivity, and layout/visual details into consistent representations.
  * Introduced a Union-Find utility to support reliable connectivity analysis during canonicalization.
  * In browser environments, the project canonicalization function is available globally for direct use.

### CODE REVIEW

File: v1/src/simulator/src/data/canonical.ts [hunk 1]
```diff
@@ -0,0 +1,975 @@
+import { circuitElementList } from "../metadata";
```
[the remark this answers is not in the document]
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts:1]: This is intentionally placed in the v1 folder to safely test the experimental pipeline separately without affecting the stable v0

File: v1/src/simulator/src/data/canonical.ts [hunk 2]
```diff
@@ -0,0 +1,975 @@
...
+    code: string;
+    subCircuitScopeIds: string[];
+  };
+};
+
+type CanonicalProject = {
+  formatVersion: "v1";
+  canonicalHash: string;
+  circuits: Record<number, CanonicalScope>;
+};
```
[the remark this answers is not in the document]
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts:160]: Scope.id is set in the constructor as Math.floor(Math.random() * 100000000000 + 1) — always a number. The only other way it's set is if (id) scope.id = id when loading — if that id is truthy, Number() of it works, so running into NaN in practice is nearly impossible

File: v1/src/simulator/src/data/canonical.ts [hunk 3]
```diff
@@ -0,0 +1,975 @@
...
+  }
+
+  const wlColours = wlFingerprint(components);
+  for (let i = 0; i < components.length; i++) {
+    const comp = components[i];
+    fpMap.set(comp, `${fpMap.get(comp)}|${wlColours.get(i) ?? ""}`);
+  }
+
+  components.sort((a, b) => naturalCompare(fpMap.get(a)!, fpMap.get(b)!));
+}
```
[the remark this answers is not in the document]
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts:534]: Claim 1: For two components to have identical WL colours despite different properties/state, they must share the same type, bitWidth, port structure, and WL-neighbourhood fingerprint. In that case they are graph-isomorphic peers — swapping their IDs produces a  semantically identical circuit. The hash may differ due to JSON key ordering in the serialized output, but the circuit behaviour is unchanged.

Claim 2: Object.entries() on _connections always iterates in insertion order, and _connections is built deterministically from customSave().nodes, which every component constructs with fixed key ordering. There is no source of indeterminism

File: v1/src/simulator/src/data/canonical.ts [hunk 4]
```diff
@@ -0,0 +1,975 @@
...
+export async function canonicaliseProject(
+  scopeOrScopes: CVScope | CVScope[],
+): Promise<CanonicalProject> {
+  const scopes = Array.isArray(scopeOrScopes) ? scopeOrScopes : [scopeOrScopes];
+  const circuits: Record<number, CanonicalScope> = {};
+  const circuitHashes: string[] = [];
+
+  for (const scope of scopes) {
+    if (!scope || !scope.allNodes) continue;
+    const circuitId = Number(scope.id);
```
Comment (aryanndwi123) [inline v1/src/simulator/src/data/canonical.ts]: `CVScope.id` is typed as `id?: number`, so `Number(undefined)` returns `NaN`. When used as an object key, `NaN` is silently coerced to the string `"NaN"`, meaning any two scopes without an `id` will overwrite each other in `circuits` with no error or warning.

File: v1/src/simulator/src/data/canonical.ts [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
+}
+
+// Typed window extension — eliminates unsafe casting / implicit any on window
+declare global {
+  interface Window {
+    canonicaliseProject?: typeof canonicaliseProject;
+  }
+}
+if (typeof window !== "undefined") {
+  window.canonicaliseProject = canonicaliseProject;
```
Comment (aryanndwi123) [inline v1/src/simulator/src/data/canonical.ts:985]: PR #1093 exposed `window.canonicaliseScope`. This PR replaces it with only `window.canonicaliseProject`, removing the previous export without notice. Anyone relying on `window.canonicaliseScope`  including the console snippet in #1093's own PR description  will silently get `undefined` at runtime.

File: v1/src/simulator/src/data/canonical.ts [hunk 6]
```diff
@@ -0,0 +1,975 @@
...
+  scopeOrScopes: CVScope | CVScope[],
+): Promise<CanonicalProject> {
+  const scopes = Array.isArray(scopeOrScopes) ? scopeOrScopes : [scopeOrScopes];
+  const circuits: Record<number, CanonicalScope> = {};
+  const circuitHashes: string[] = [];
+
+  for (const scope of scopes) {
+    if (!scope || !scope.allNodes) continue;
+    const circuitId = Number(scope.id);
+    const circuit = await canonicaliseScope(scope);
```
Comment (aryanndwi123) [inline v1/src/simulator/src/data/canonical.ts]: Each scope's SHA-256 hashing is awaited one at a time. These calls are fully independent of each other, so this is leaving parallelism on the table that grows linearly with circuit count.

**Review by aryanndwi123 — COMMENTED** [review]
Good foundation for multi-circuit support, the hash-sorting approach for order-independence is correct, a few issues need addressing before merge

File: v1/src/simulator/src/data/canonical.ts (hunk 4 above)
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts]: Agreed. I will fix this by throwing an error if the id is undefined. I am not including a fallback id like -1, as it can lead to non-determinism.

File: v1/src/simulator/src/data/canonical.ts (hunk 5 above)
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts:985]: `window.canonicaliseProject` is enough, and `window.canonicaliseScope` is not needed. Btw, the Typed window extension is for testing purposes only and will be removed afterwards.

File: v1/src/simulator/src/data/canonical.ts (hunk 6 above)
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts]: That's a great point! Adding Parallelism will be time-efficient!

**Comment by Harkeerat24** [comment]
@aryanndwi123 @JoshVarga @tachyons PTAL

**Comment by tachyons** [comment]
Please check the PR diff before requesting review and ensure that there aren't any unrelated changes

**Comment by Harkeerat24** [comment]
> Please check the PR diff before requesting review and ensure that there aren't any unrelated changes

I am extremely sorry, @tachyons. I didn't recheck the diff properly before committing. I have fixed it now, and I'll make sure to double-check my changes carefully before requesting reviews in the future. Sorry about that!

**Comment by aryanndwi123** [comment]
Hey @Harkeerat24 there are some merge conflicts in here

**Comment by Harkeerat24** [comment]
> Hey @Harkeerat24 there are some merge conflicts in here

@aryanndwi123 This pr is not required now, as #1095, which is merged, was built on top of this. Hence, #1094 and #1093 are both not needed.

**Comment by aryanndwi123** [comment]
> > Hey @Harkeerat24 there are some merge conflicts in here
> 
> @aryanndwi123 This pr is not required now, as #1095, which is merged, was built on top of this. Hence, #1094 and #1093 are both not needed.

Should we close both the PRs?

**Comment by Harkeerat24** [comment]
#1095 fixes this, closing this PR

**Comment by Harkeerat24** [comment]
> > > Hey @Harkeerat24 there are some merge conflicts in here
> > 
> > 
> > @aryanndwi123 This pr is not required now, as #1095, which is merged, was built on top of this. Hence, #1094 and #1093 are both not needed.
> 
> Should we close both the PRs?

DONE!
